### PR TITLE
Fix `segmentation.MeanIoU`

### LINKED ISSUE #2558: segmentation.MeanIoU is wrong
## 🐛 Bug
MeanIoU scored 56(!) over validation dataset
image

Let's look at the source code:
https://github.com/Lightning-AI/torchmetrics/blob/596ed09b18c3d6786cb094c87da97838228461f3/src/torchmetrics/segmentation/mean_iou.py#L109

```py
def update(self, preds: Tensor, target: Tensor) -> None:
    """Update the state with the new data."""
    intersection, union = _mean_iou_update(preds, target, self.num_classes, self.include_background)
    score = _mean_iou_compute(intersection, union, per_class=self.per_class)
    self.score += score.mean(0) if self.per_class else score.mean()

def compute(self) -> Tensor:
    """Update the state with the new data."""
    return self.score  # / self.num_batches
```

There are several issues there:
- `self.score` is accumulated with each call of `update` method.
- `compute` method just returns the accumulated score
- description of `compute ` method is copypasted from `update` method
- division by `self.num_batches` is commented
- `num_batches` is definded on class level and not used anywhere else

Obviously, that code was neither reviewed nor tested, but somehow was released.

### To Reproduce

Call `metric.update(y_hat, y)` in `validation_step`
Log the metric in `on_validation_epoch_end`

### Expected behavior

MeanIoU computes correct value in [0, 1] range.

### Environment

- TorchMetrics 1.4.0.post0 from pip
- Python 3.11.9, PyTorch 2.3.0+cu121
- pytorch-lightning 2.2.4
- Windows 11

### PR DESCRIPTION
**Pull request by vkinakh**
## What does this PR do?

Fixes #2558 
Fixes a few typos

<details>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?
</details>

This PR fixes the `segmentation.MeanIoU` metric. 

# Issues
- when the metric is updated via calling `self.update`, the `self.score` is accumulated with each call. Then when `self.compute` is called, accumulated metric is returned
- when the metric is updated via calling `self.forward`, metric works correct

It is happens because when `MeanIoU` is updated via `self.forward`, it calls `self._reduce_states` method, which updates score using the following formula:

`reduced = ((self._update_count - 1) * global_state + local_state).float() / self._update_count`

where `global_state` is the score accumulated over previous steps and `local_state` is the score on current batch.

# Solution

To solve this issue:
- use sum reduce function for `self.score`
- add state `num_batches` to keep number of processed batches
- add increment of `num_batches` in every `self.update` call
- in `self.compute` return sum of scores divided by number of processed batches

----
📚 Documentation preview 📚: https://torchmetrics--2698.org.readthedocs.build/en/2698/

### CODE REVIEW

**Comment by vkinakh** [comment]
Here is an example code to reproduce the issue

```python
import
from torchmetrics.segmentation import MeanIoU

```

**Comment by vkinakh** [comment]
Here

File: src/torchmetrics/segmentation/mean_iou.py [hunk 1]
```diff
@@ -120,10 +121,11 @@ def update(self, preds: Tensor, target: Tensor) -> None:
         )
         score = _mean_iou_compute(intersection, union, per_class=self.per_class)
         self.score += score.mean(0) if self.per_class else score.mean()
+        self.num_batches += 1
 
     def compute(self) -> Tensor:
-        """Update the state with the new data."""
-        return self.score  # / self.num_batches
```
Comment (Borda) [inline src/torchmetrics/segmentation/mean_iou.py:126]: @SkafteNicki any idea why this was commented?

**Review by Borda — COMMENTED** [review]
great work, can we please add a test to prevent this issue in the future?

File: src/torchmetrics/segmentation/mean_iou.py (hunk 1 above)
Comment (vkinakh) [reply src/torchmetrics/segmentation/mean_iou.py:126]: I think the best would be to add a testcase to `MetricTester`, that checks if the return of `self.compute` is the same, when class is updated using `self.forward` and `self.update`

**Comment by vkinakh** [comment]
I have also noticed that names and descriptions in test for `GeneralizedDiceScore` are just copy-paste of `MeanIoU` tests. Tests are correct, but names and descriptions are incorrect and misleading

Should I create a new issue?

https://github.com/Lightning-AI/torchmetrics/blob/24e99c5d80674f4145c0b21027eb9fe579e36bb6/tests/unittests/segmentation/test_generalized_dice_score.py#L66-L108

**Comment by Borda** [comment]
> I have also noticed that names and descriptions in test for `GeneralizedDiceScore` are just copy-paste of `MeanIoU` tests.

that was a typo, fixing it in #2709

**Comment by SkafteNicki** [comment]
Added generalized testing that `forward/update` works as expected (comment https://github.com/Lightning-AI/torchmetrics/pull/2698#discussion_r1727149727) in commit https://github.com/Lightning-AI/torchmetrics/pull/2698/commits/3d3d7b5a6924119bb16293d9dc22e7fdd54eb9dd. Hopefully, this check should pass for all other metrics, else lets move this to another PR.

**Review by SkafteNicki — APPROVED** [review]

**Comment by vkinakh** [comment]
> Added generalized testing that `forward/update` works as expected (comment [#2698 (comment)](https://github.com/Lightning-AI/torchmetrics/pull/2698#discussion_r1727149727)) in commit [3d3d7b5](https://github.com/Lightning-AI/torchmetrics/commit/3d3d7b5a6924119bb16293d9dc22e7fdd54eb9dd). Hopefully, this check should pass for all other metrics, else lets move this to another PR.

I suggest adding a test case to verify the metric aggregation behavior: the aggregated metric of a batch with 2*N predictions should match the aggregated metrics from two separate updates, each with N predictions. Are there any metrics that are not expected to behave this way?

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**